lib/Browser.js

```diff
@@ -31,6 +31,58 @@ class Browser extends EventEmitter {
     this._screenshotTaskQueue = new TaskQueue();
     this._connection = connection;
     this._closeCallback = closeCallback || new Function();
+    /** @type {Map<string, Target>} */
+    this._targets = new Map();
+    this._connection.on('Target.targetCreated', this._targetCreated.bind(this));
+    this._connection.on('Target.targetDestroyed', this._targetDestroyed.bind(this));
+    this._connection.on('Target.targetInfoChanged', this._targetInfoChanged.bind(this));
+  }
+
+  /**
+   * @param {!Connection} connection
+   * @param {boolean} ignoreHTTPSErrors
+   * @param {function()=} closeCallback
+   */
+  static async create(connection, ignoreHTTPSErrors, closeCallback) {
+    const browser =  new Browser(connection, ignoreHTTPSErrors, closeCallback);
+    await connection.send('Target.setDiscoverTargets', {discover: true});
+    return browser;
+  }
+
+  /**
+   * @param {{targetInfo: Target.TargetInfo}} param0
+   */
+  async _targetCreated({targetInfo}) {
+    const target = await this._ensureTargetForInfo(targetInfo);
+    this.emit(Browser.Events.TargetCreated, target);
+  }
+
+  /**
+   * @param {{targetId: string}} param0
+   */
+  async _targetDestroyed({targetId}) {
+    const target = this._targets.get(targetId);
+    if (!target)
+      return;
+    this._targets.delete(targetId);
+    this.emit(Browser.Events.TargetDestroyed, target);
+  }
+
+  /**
+   * @param {{targetInfo: Target.TargetInfo}} param0
+   */
+  async _targetInfoChanged({targetInfo}) {
+    this._ensureTargetForInfo(targetInfo)._targetInfo = targetInfo;
```

Comment:
do we want to emit 'targetChanged' here?

Reply:
Let's add in a later patch.